### PR TITLE
 I8 MFMA 16x16x64 and 32x32x32 instruction added

### PR DESCRIPTION
**Pull request by adityalj**
GFX950 supports I8 MFMA of size 16x16x64 and 32x32x32.

### CODE REVIEW

File: tensilelite/Tensile/SolutionStructs/Solution.py [hunk 1]
```diff
@@ -924,12 +924,12 @@ def isDirectToLdsDoable(state, tc, isaInfoMap, printRejectionReason: bool):
     # ToDo: Review def of lrvw and this check
     # DTL + LocalReadVectorWidth > MIInputPerThread does not work
     # Need support for TailLoop
-    if not state["ProblemType"]["Sparse"] and (not isaInfoMap[isa].asmCaps["HasMFMA_f8f6f4"] or state["MatrixInstK"] <= 32):
+    if not state["ProblemType"]["Sparse"] and (not isaInfoMap[isa].asmCaps["HasMFMA_f8f6f4"] or not(state["ProblemType"]["DataType"].is8bitFloat() and (state["MatrixInstK"] == 64 or state["MatrixInstK"] == 128))):
```
Comment (msujon-AMD) [inline tensilelite/Tensile/SolutionStructs/Solution.py]: We don't need HasMFMA_f8f6f4 checking anymore since we are checking F8 with K=64 and 128 explicitly.

File: tensilelite/Tensile/SolutionStructs/Solution.py [hunk 2]
```diff
@@ -924,12 +924,12 @@ def isDirectToLdsDoable(state, tc, isaInfoMap, printRejectionReason: bool):
...
       if state["LocalReadVectorWidth"] > state["MIInputPerThread"]:
         reject(state, printRejectionReason, "DirectToLds does not work with LocalReadVectorWidth > MIInputPerThread")
         return False
 
-    if not state["ProblemType"]["Sparse"] and (not isaInfoMap[isa].asmCaps["HasMFMA_f8f6f4"] or state["MatrixInstK"] <= 32):
+    if not state["ProblemType"]["Sparse"] and (not isaInfoMap[isa].asmCaps["HasMFMA_f8f6f4"] or not(state["ProblemType"]["DataType"].is8bitFloat() and (state["MatrixInstK"] == 64 or state["MatrixInstK"] == 128))):
```
Comment (msujon-AMD) [inline tensilelite/Tensile/SolutionStructs/Solution.py]: same as previous comment

File: tensilelite/Tensile/SolutionStructs/Solution.py [hunk 3]
```diff
@@ -1774,7 +1774,7 @@ def calcLdsNumBytes(ldsPadA: int, LdsBlockSizePerPadA: int, ldsPadB: int, LdsBlo
           if state["ProblemType"]["Sparse"] and state["MIInputPerThread"] * state["ProblemType"]["DataType"].numBytes() > 16:
             if state["LocalReadVectorWidth"] < state["MIInputPerThread"] // 2:
               reject(state, printRejectionReason, "LocalReadVectorWidth < %u" %(state["MIInputPerThread"] // 2))
-          elif not state["ProblemType"]["Sparse"] and (not isaInfoMap[isa].asmCaps["HasMFMA_f8f6f4"] or state["MatrixInstK"] <= 32):
+          elif not state["ProblemType"]["Sparse"] and (not isaInfoMap[isa].asmCaps["HasMFMA_f8f6f4"] or not(state["ProblemType"]["DataType"].is8bitFloat() and (state["MatrixInstK"] == 64 or state["MatrixInstK"] == 128))):
```
Comment (msujon-AMD) [inline tensilelite/Tensile/SolutionStructs/Solution.py]: same as previous comment

File: tensilelite/Tensile/SolutionStructs/Solution.py [hunk 4]
```diff
@@ -3039,7 +3039,7 @@ def calcEpilogueTurns(factorDims: List) -> int:
       # Multiple = WLR-size / input-size = how many iters could be covered by one WLR ?
       wlrMultiple = state["LocalReadVectorWidth"]//state["MIInputPerThread"]
       # NOTE: wlrmultiple can be 0 for new MFMA
-      if not state["ProblemType"]["Sparse"] and (not isaInfoMap[isa].asmCaps["HasMFMA_f8f6f4"] or state["MatrixInstK"] <= 32):
+      if not state["ProblemType"]["Sparse"] and (not isaInfoMap[isa].asmCaps["HasMFMA_f8f6f4"] or not(state["ProblemType"]["DataType"].is8bitFloat() and (state["MatrixInstK"] == 64 or state["MatrixInstK"] == 128))):
```
Comment (msujon-AMD) [inline tensilelite/Tensile/SolutionStructs/Solution.py]: same as previous comment

File: tensilelite/Tensile/Tests/common/gemm/i8_gsu.yaml [hunk 5]
```diff
@@ -70,6 +70,104 @@ BenchmarkProblems:
         - ActivationArgs:
           - [Enum: none]
 
+  #######################################
```
Comment (msujon-AMD) [inline tensilelite/Tensile/Tests/common/gemm/i8_gsu.yaml]: We should not change the existing test yaml with new MFMA since it will fail tests on gfx942 and other arch. Let's create a separate test yaml inside gfx950 directory

**Review by msujon-AMD — CHANGES_REQUESTED** [review]

File: tensilelite/Tensile/Tests/common/gemm/gfx950/i8_gsu_gfx950.yaml [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        - 1LDSBuffer: [-1]
+        - DirectToVgprSparseMetadata: [0]
+        - WorkGroupMapping: [18]
+        - StoreVectorWidth: [1]
+      BenchmarkForkParameters:
+      JoinParameters:
+      BenchmarkJoinParameters:
+      BenchmarkFinalParameters:
+        - ProblemSizes:
+          - Exact: [256, 256, 1, 256] # classic format
```
Comment (msujon-AMD) [inline tensilelite/Tensile/Tests/common/gemm/gfx950/i8_gsu_gfx950.yaml:68]: Let's add more sizes which exercises tailloop and edge cases. We can also try prime numbers as well.

File: tensilelite/Tensile/Tests/common/gemm/gfx950/i8_gsu_gfx950.yaml [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+BenchmarkProblems:
+  #######################################
+  # NT - standard
+  #######################################
+  -
+    - # ProblemType
+      OperationType: GEMM
+      DataType: i8
+      DestDataType: i8
```
Comment (msujon-AMD) [inline tensilelite/Tensile/Tests/common/gemm/gfx950/i8_gsu_gfx950.yaml:30]: Let's have DestDataType as INT32 as well.

File: tensilelite/Tensile/Tests/common/gemm/gfx950/i8_gsu_gfx950.yaml [hunk 8]
```diff
@@ -0,0 +1,120 @@
+TestParameters:
+  marks: [skip-gfx940, skip-gfx941, skip-gfx942, skip-gfx1100, skip-gfx1101, skip-gfx1102, skip-gfx1200, skip-gfx1201] # not supported by arch
```
Comment (msujon-AMD) [inline tensilelite/Tensile/Tests/common/gemm/gfx950/i8_gsu_gfx950.yaml]: we need to skip this tests on gfx90a as well.

File: clients/gtest/matmul_gtest.yaml [hunk 9]
```diff
@@ -443,6 +443,38 @@ Tests:
   unit_check: 1
   gpu_arch: '950'
 
+- name: matmul_gemm_i8_dst_i32
```
Comment (msujon-AMD) [inline clients/gtest/matmul_gtest.yaml]: do we need separate test case for gfx950 or we can use existing one?

**Review by msujon-AMD — DISMISSED** [review]

File: clients/gtest/matmul_gtest.yaml (hunk 9 above)
Comment (adityalj) [reply clients/gtest/matmul_gtest.yaml]: You are right. I modified previous test for 90a to include 950.

**Comment by msujon-AMD** [comment]
LGTM. The liblogic is a placeholder and we will replace it after tuning.

**Review by msujon-AMD — DISMISSED** [review]

File: tensilelite/Tensile/Components/LocalRead.py [hunk 10]
```diff
@@ -525,7 +525,7 @@ def __call__(self, writer, kernel, bufferIdx, iui, epsi, tP):
                                 incOffset = 0
                                 midIdx = numReadsPerUnroll // 2
                                 if rIdx >= midIdx:
-                                    if kernel["MatrixInstK"] > 32:
+                                    if  kernel["ProblemType"]["DataType"].is8bitFloat() and kernel["MatrixInstK"] > 32:
```
Comment (KKyang) [inline tensilelite/Tensile/Components/LocalRead.py]: These hard coding lines are stacking more and more.
In L524 you have HasMFMA_f8f6f4, and suddenly the condition in L528 becomes if ``is8bitfloat()`` and MIK > 32. Is MIK > 32 bounded to 8bit float?

File: tensilelite/Tensile/Components/LraTileAssignment.py [hunk 11]
```diff
@@ -207,7 +207,7 @@ def LraTileAssignmentCode(self, writer, kernel, tP, tReg, kReg, tmpVgprRes, divi
               isSparseTrack = (kernel["ProblemType"]["Sparse"] == 2 and tP["isB"]) or (kernel["ProblemType"]["Sparse"] == 1 and tP["isA"]) or tP["isM"]
               strideK      = (inputPerThread if umlds else (mt + LdsPad) * inputPerThread) * (2 if isSparseTrack and kernel["MIInputPerThread%s"%tc] >  inputPerThread else 1)
         #special case for new F8 MFMA
-        elif kernel["MatrixInstK"] > 32:
+        elif  kernel["ProblemType"]["DataType"].is8bitFloat() and kernel["MatrixInstK"] > 32:
```
Comment (KKyang) [inline tensilelite/Tensile/Components/LraTileAssignment.py:211]: Same here, see the comment above.

File: tensilelite/Tensile/SolutionStructs/Solution.py [hunk 12]
```diff
@@ -925,12 +925,12 @@ def isDirectToLdsDoable(state, tc, isaInfoMap, printRejectionReason: bool):
     # ToDo: Review def of lrvw and this check
     # DTL + LocalReadVectorWidth > MIInputPerThread does not work
     # Need support for TailLoop
-    if not state["ProblemType"]["Sparse"] and (not isaInfoMap[isa].asmCaps["HasMFMA_f8f6f4"] or state["MatrixInstK"] <= 32):
+    if not state["ProblemType"]["Sparse"] and not(state["ProblemType"]["DataType"].is8bitFloat() and (state["MatrixInstK"] == 64 or state["MatrixInstK"] == 128)):
```
Comment (KKyang) [inline tensilelite/Tensile/SolutionStructs/Solution.py:927]: Are these temporarily disabled or will be never supported? While changing the condition from ``HasMFMA_f8f6f4`` to ``is8bitfloat()`` also affects other platform. Is this the condition you want?

**Review by msujon-AMD — DISMISSED** [review]

File: tensilelite/Tensile/SolutionStructs/Solution.py (hunk 12 above)
Comment (adityalj) [reply tensilelite/Tensile/SolutionStructs/Solution.py:927]: Only new F8 can have MI_K=64 or 128. I think the F8F6F4 checking is redundant.  We have enabled smaller LRVW for new F8 instruction to experiment some optimizations. but it is only applicable to new F8 MFMA

File: tensilelite/Tensile/Components/LraTileAssignment.py (hunk 11 above)
Comment (adityalj) [reply tensilelite/Tensile/Components/LraTileAssignment.py:211]: the update in StrideK is only needed for new F8 MFMA in gfx950 having MI_K >32 since the register layout is different

File: tensilelite/Tensile/Components/LocalRead.py (hunk 10 above)
Comment (adityalj) [reply tensilelite/Tensile/Components/LocalRead.py]: This condition is true only for new MFMAs where datatype is F8 and MI_K>32 as register layout is different. As per your suggestions we have changed the condition and reduced hardcoding.

**Review by msujon-AMD — APPROVED** [review]

**Review by jichangjichang — APPROVED** [review]
